identity-account-storage/src/types/key_location.rs

```diff
@@ -6,60 +6,105 @@ use core::fmt::Display;
 use core::fmt::Formatter;
 use core::fmt::Result;
 use identity_core::common::Fragment;
+use identity_core::crypto::KeyType;
+use identity_did::verification::MethodData;
 use identity_did::verification::MethodType;
-use serde::Deserialize;
-use serde::Serialize;
-
-use crate::types::Generation;
+use identity_iota_core::document::IotaVerificationMethod;
+use rand::distributions::Alphanumeric;
+use rand::prelude::ThreadRng;
+use rand::Rng;
+use seahash::SeaHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
 
 /// The storage location of a verification method key.
-#[derive(Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
+///
+/// A key is uniquely identified by the fragment and a hash of its public key.
+/// Importantly, the fragment alone is insufficient to represent the storage location.
+/// For example, when rotating a key, there will be two keys in storage for the
+/// same identity with the same fragment. The `key_hash` disambiguates the keys in
+/// situations like these.
+///
+/// The string representation of that location can be obtained via `to_string`.
+#[derive(Clone, Hash, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 pub struct KeyLocation {
-  method: MethodType,
+  key_type: KeyType,
   fragment: Fragment,
-  generation: Generation,
+  key_hash: u64,
 }
 
 impl KeyLocation {
-  /// Creates a new `KeyLocation`.
-  pub fn new(method: MethodType, fragment: String, generation: Generation) -> Self {
+  pub fn new(key_type: KeyType, fragment: String, public_key: &[u8]) -> Self {
+    let mut hasher = SeaHasher::new();
+    public_key.hash(&mut hasher);
+    let key_hash = hasher.finish();
+
     Self {
-      method,
+      key_type,
       fragment: Fragment::new(fragment),
-
-      generation,
+      key_hash,
     }
   }
 
-  /// Returns the method type of the key location.
-  pub fn method(&self) -> MethodType {
-    self.method
+  /// Generates a random location for a key of the given [`KeyType`].
+  pub fn random(key_type: KeyType) -> Self {
+    let mut thread_rng: ThreadRng = rand::thread_rng();
+    let fragment: String = (&mut thread_rng)
+      .sample_iter(Alphanumeric)
+      .take(32)
+      .map(char::from)
+      .collect();
+    let key_hash: u64 = (&mut thread_rng).gen();
+
+    Self {
+      key_type,
+      fragment: Fragment::new(fragment),
+      key_hash,
+    }
   }
```

Comment:
You're right, probably not necessary. That also nicely enables us to feature-gate the `rand` crate to the `stronghold` feature, so no Wasm issues.

Reply:
There is a location where I used `rand` when `tesmode` is set to true in `Account`. There wasn't an elegant way to feature-gate it, but maybe we can go with an unelegant solution. It wasn't exposed to Wasm in my situation since `testmode` is internal-only.